test/Rings-conformance-tests.jl

```diff
@@ -0,0 +1,222 @@
+            for i in 1:reps
+               a = rand(elems)
+               b = rand(elems)
+               A = deepcopy(a)
+               B = deepcopy(b)
+               @test iszero(b) || equality(divexact_left(b*a, b), a)
+               @test iszero(b) || equality(divexact_right(a*b, b), a)
+               @test A == a
+               @test B == b
+            end
+         end
+      end
+
+      @testset "Unsafe ring operators" begin
+         for i in 1:reps
+            a = rand(elems)
+            b = rand(elems)
+            c = rand(elems)
+            A = deepcopy(a)
+            B = deepcopy(b)
+            C = deepcopy(c)
+
+            x = deepcopy(a)
+            @test iszero(zero!(x))
+
+            ab = a*b
+            x = R()
+            @test mul!(x, a, b) == ab
+            x = deepcopy(b)
+            @test mul!(x, a, b) == ab
+            x = deepcopy(a)
+            @test mul!(x, x, b) == ab
+            x = deepcopy(b)
+            @test mul!(x, a, x) == ab
+
+            ab = a+b
+            x = R()
+            @test add!(x, a, b) == ab
+            x = deepcopy(b)
+            @test add!(x, a, b) == ab
+            x = deepcopy(a)
+            @test add!(x, x, b) == ab
+            x = deepcopy(b)
+            @test add!(x, a, x) == ab
+            x = deepcopy(a)
+            @test addeq!(x, b) == ab
+
+            # hmmm, doc says three args but describes four args 
+            # BigFloat may fuse, so isapprox
+            t = R()
+            x = deepcopy(a)
+            @test equality(addmul!(x, b, c, t), a+b*c)
+            x = deepcopy(a)
+            @test equality(addmul!(x, x, c, t), a+a*c)
+            x = deepcopy(a)
+            @test equality(addmul!(x, b, x, t), a+b*a)
+            x = deepcopy(a)
+            @test equality(addmul!(x, x, x, t), a+a*a)
+
+            @test A == a
+            @test B == b
+            @test C == c
+         end         
+      end
+   end
+end
+
+
+function test_Ring_interface(
+   R::AbstractAlgebra.Ring,
+   example_elements::Vector{T}
+) where T <: NCRingElement
+
+   test_NCRing_interface(R, example_elements)
+
+
+   elems = deepcopy(example_elements)
+   append!(elems, T[R(0), R(1), R(-2), R(BigInt(0)), R(BigInt(1)),
+                    R(BigInt(-2))#=, R(BigInt(3)^100)=#])
+
+   if !iszero(one(R)) && isdomain_type(T)
+      @testset "Basic functionality for commutative rings only" begin
+         for i in 1:reps
+            a = rand(elems)
+            b = rand(elems)
+            A = deepcopy(a)
+            B = deepcopy(b)
+            @test isone(inv(one(R)))
+            @test a*b == b*a
+            @test iszero(b) || divexact(b*a, b) == a
+            @test A == a
+            @test B == b
+         end
+      end
+   end
+end
+
+
+#InexactError on R(BigInt(3) ^ 100)
+#failure on iszero(b) || divexact(b * a, b) == a
+#test_Ring_interface(zz, Int[11, 2^30, 2^40])
+
+test_Ring_interface(ZZ, BigInt[11])
```

Comment:
I would recommend to move the invocations of `test_Ring_interface` to another file, ideally to the files containing the other tests for the ring(s) in question. This way other packages can load this file to be able to use `test_Ring_interface`.

Reply:
will do ...